### PR TITLE
fix: use remaining parameters as query params when building URLs

### LINKED ISSUE #2132: Support sending optional query params in ApiHttpService's setEndpointUrl method
## This is a...

<pre><code>
[X] Feature request
[ ] Regression (a behavior that used to work and stopped working in a new release)
[ ] Bug report  
[ ] Documentation issue or request
</code></pre>

## The problem
There are APIs which support optional query parameters.  For example all entity listing APIs support pagination control via query parameters.  

We need a way the UI pass these to the APIs when needed.  Keep in mind the query parameters keys and values should be URL encoded and that query keys are allowed to repeat.  Like:

/api/somepath?id=1&id=2&id=3

### PR DESCRIPTION
**Pull request by gashcrumb**
fixes #2132

### CODE REVIEW

File: app/ui/src/app/api/providers/api-http-provider.service.ts [hunk 1]
```diff
@@ -165,9 +165,40 @@ export class ApiHttpProviderService extends ApiHttpService {
...
+      const value = paramsMap[key];
+      // avoid passing any url param as 'undefined'
+      if (!value) {
+        return undefined;
+      }
+      // expand an array value in the param map to multiple query params
+      if (typeof value === 'string') {
+        return key + '=' + value;
+      } else if (typeof value === 'object' && value.map) {
+        return value.map(val => key + '=' + val).join('&');
```
Comment (chirino) [inline app/ui/src/app/api/providers/api-http-provider.service.ts]: we should probably URL encode key and values

File: app/ui/src/app/api/providers/api-http-provider.service.ts [hunk 2]
```diff
@@ -165,9 +165,40 @@ export class ApiHttpProviderService extends ApiHttpService {
...
   private replaceNameMatches(endpointTemplate: string, paramsMap: StringMap<any>): string {
-    return endpointTemplate.replace(/\{(\D*?)\}/g, (fullMatch, ...matchGroups) => {
-      return paramsMap[matchGroups[0].trim()];
+    const matched = [];
+    const answer = endpointTemplate.replace(/\{(\D*?)\}/g, (fullMatch, ...matchGroups) => {
+      const match = matchGroups[0].trim();
+      matched.push(match);
+      return paramsMap[match];
     });
+    // Turn any remaining parameters into query params
```
Comment (chirino) [inline app/ui/src/app/api/providers/api-http-provider.service.ts]: We could probably use https://angular.io/api/common/http/HttpParams to make sure we properly encode the query string.

Comment (gashcrumb) [reply app/ui/src/app/api/providers/api-http-provider.service.ts]: yeah, good call, will add that, thanks!

Comment (deeleman) [reply app/ui/src/app/api/providers/api-http-provider.service.ts]: Perhaps you will want to use the `HttpUrlEncodingCodec` class instead (it's not injectable, you must instantiate it with `new`). It features both `encodeKey()` and `encodeValue()` methods.

**Pull request merged**